server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java

```diff
@@ -965,122 +934,102 @@ public Map<String, List<String>> getHeaders() {
-            @Override
-            public void dispatchRequest(RestRequest request, RestChannel channel, ThreadContext threadContext) {
-                inDispatch.countDown();
-            }
-
-            @Override
-            public void dispatchBadRequest(RestChannel channel, ThreadContext threadContext, Throwable cause) {
-                channel.sendResponse(emptyResponse(RestStatus.BAD_REQUEST));
-            }
-        })) {
-            Loggers.addAppender(mockLogger, appender);
-            appender.start();
+        var grace = 10;
+        TestHttpChannel httpChannel = new TestHttpChannel();
+        try (
+            var timeout = LogExpectation.expectTimeout(grace);
+            TestHttpServerTransport transport = new TestHttpServerTransport(gracePeriod(grace))
+        ) {
 
-            appender.addExpectation(
-                new MockLogAppender.SeenEventExpectation(
-                    "message",
-                    AbstractHttpServerTransport.class.getName(),
-                    Level.WARN,
-                    "timed out while waiting [10]ms for clients to close connections"
-                )
-            );
+            httpChannel.blockSendResponse();
+            var inResponse = httpChannel.notifyInSendResponse();
 
-            transport.bindServer();
-            final TestHttpRequest httpRequest = new TestHttpRequest(HttpRequest.HttpVersion.HTTP_1_1, RestRequest.Method.GET, "/");
-            TestHttpChannel httpChannel = new TestHttpChannel();
             transport.serverAcceptedChannel(httpChannel);
             new Thread(
-                () -> transport.incomingRequest(httpRequest, httpChannel),
+                () -> transport.incomingRequest(testHttpRequest(), httpChannel),
                 "testStopForceClosesConnectionDuringRequest -> incomingRequest"
             ).start();
-            inDispatch.await();
+
+            inResponse.await();
+
             assertTrue(httpChannel.isOpen());
             transport.doStop();
+
             assertFalse(httpChannel.isOpen());
             assertFalse(transport.testHttpServerChannel.isOpen());
-            assertThat(httpChannel.responses, hasSize(0));
+            assertTrue(httpChannel.noResponses());
+
             // ensure we timed out waiting for connections to close naturally
-            appender.assertAllExpectationsMatched();
+            timeout.assertExpectationsMatched();
         } finally {
-            appender.stop();
-            Loggers.removeAppender(mockLogger, appender);
+            // unblock request thread
+            httpChannel.allowSendResponse();
         }
     }
 
-    public void testStopClosesChannelAfterRequest() {
-        try (TestHttpServerTransport transport = new TestHttpServerTransport(gracePeriod(100))) {
-            transport.bindServer();
+    public void testStopClosesChannelAfterRequest() throws Exception {
+        var grace = 100;
```

Comment:
100ms seems possible to hit if the test doesn't get cpu cycles for a while. It seems like there are essentially 2 grace periods for test we care about, incredible short (so that we hit the timeout) and incredibly long (so we don't hit the timeout). Why not use 1ms and (effectively) infinity (could be eg 20 seconds in case of problem so we don't wait forever for the test to fail) for these values?

Reply:
`testForceClosesOpenChannels` needs to wait the entire time, so there is also an intermediate period.

Reply:
I've added a comment in `testForceClosesOpenChannels` and added `LONG_GRACE_PERIOD_MS` and `SHORT_GRACE_PERIOD_MS` constants for use in other tests.